constraint/solver/gkrgates/registry.go

```diff
@@ -35,78 +37,155 @@ type registerSettings struct {
+	return func(settings *registerSettings) error {
+		if settings.solvableVar != -1 {
+			return fmt.Errorf("solvable variable already set to %d", settings.solvableVar)
+		}
+		if settings.noSolvableVarVerification {
+			return errors.New("solvable variable already set to NONE")
+		}
 		settings.noSolvableVarVerification = true
 		settings.solvableVar = solvableVar
+		return nil
 	}
 }
 
 // WithNoSolvableVar sets the gate as having no variable whose value can be uniquely determined from that of the other variables along with the gate's output.
 // RegisterGate will not check the correctness of this claim.
-func WithNoSolvableVar() registerOption {
-	return func(settings *registerSettings) {
+func WithNoSolvableVar() RegisterOption {
+	return func(settings *registerSettings) error {
+		if settings.solvableVar != -1 {
+			return fmt.Errorf("solvable variable already set to %d", settings.solvableVar)
+		}
+		if settings.noSolvableVarVerification {
+			return errors.New("solvable variable already set to NONE")
+		}
 		settings.solvableVar = -1
 		settings.noSolvableVarVerification = true
+		return nil
 	}
 }
 
 // WithUnverifiedDegree sets the degree of the gate. RegisterGate will not verify that the given degree is correct.
-func WithUnverifiedDegree(degree int) registerOption {
-	return func(settings *registerSettings) {
+func WithUnverifiedDegree(degree int) RegisterOption {
+	return func(settings *registerSettings) error {
+		if settings.degree != -1 {
+			return fmt.Errorf("gate degree already set to %d", settings.degree)
+		}
 		settings.noDegreeVerification = true
 		settings.degree = degree
+		return nil
 	}
 }
 
 // WithDegree sets the degree of the gate. RegisterGate will return an error if the degree is not correct.
-func WithDegree(degree int) registerOption {
-	return func(settings *registerSettings) {
+func WithDegree(degree int) RegisterOption {
+	return func(settings *registerSettings) error {
+		if settings.degree != -1 {
+			return fmt.Errorf("gate degree already set to %d", settings.degree)
+		}
 		settings.degree = degree
+		return nil
 	}
 }
 
 // WithName can be used to set a human-readable name for the gate.
-func WithName(name gkr.GateName) registerOption {
-	return func(settings *registerSettings) {
+func WithName(name gkr.GateName) RegisterOption {
+	return func(settings *registerSettings) error {
+		if name == "" {
+			return errors.New("gate name must not be empty")
+		}
+		if settings.name != "" {
+			return fmt.Errorf("gate name already set to \"%s\"", settings.name)
+		}
 		settings.name = name
+		return nil
 	}
 }
 
 // WithCurves determines which curves the gate is validated on.
 // The default is to validate on BN254.
 // This works for most gates, unless the leading coefficient is divided by
 // the curve's order, in which case the degree will be computed incorrectly.
-func WithCurves(curves ...ecc.ID) registerOption {
-	return func(settings *registerSettings) {
+func WithCurves(curves ...ecc.ID) RegisterOption {
```

Comment:
The documentation does not match the implementation. Now in the registration you set `allowedCurves` based on the option.

For me this is complex - I think the gate should know itself where it can work (if it doesn't work over all fields) and then panic.

And I think we can still use the approach where we test the degree of the gate on a random modulus instead of letting the users define it themselves. Otherwise the user has to know what is the relation between gate degree and the provided modulus, which doesn't make sense as it is used for automatic degree testing.